### PR TITLE
Add superPartition and update documentation

### PR DESCRIPTION
**Pull request by katelynqueen98**
Added superPartition.R file and .Rd file generated with Roxygen2. Updated description file with necessary packages, my information, and updated version number.

### CODE REVIEW

**Comment by malcolmbarrett** [comment]
Thanks! I'll make a deeper review as soon as I can, hopefully by early next week, if not this week. I'll take a deeper look at the check issue then. (Since this isn't a heavy maintenance package, we should probably run checks on a schedule so we catch things like this earlier!)

In the meantime, I see a few issues immediately:
1. The code style is inconsistent with the rest of the package. Please use [`snake_case` instead of `camelCase`](https://style.tidyverse.org/syntax.html#object-names). Also, please remove the [explicit `return()` statement](https://style.tidyverse.org/functions.html#return).
2. Please revert the package version. Adding this isn't a release, so we shouldn't change the version. We'll change the version when we release it to CRAN after this is merged.
3. Please remove `.DS_store`. It's not needed and shouldn't be in the PR. Consider running `usethis::git_vaccinate()` in your R session to avoid committing that and other files that are problematic. 
4. Please add tests. The package is very well-tested, and we should keep it that way. See the current tests for examples, and let me know if you have questions.

**Comment by katelynqueen98** [comment]
I made the requested updates! 

Also, in looking closer at the CMD checks, I had a similar issue when I cloned the repository and built it. I had to manually go into my Makeconf file and add the appropriate gfortran file path and dependencies to get everything to build correctly.

File: R/superPartition.R
Comment (malcolmbarrett) [inline R/superPartition.R]:
```suggestion
    for (i in seq_along(modules)) {
```

Comment (malcolmbarrett) [reply R/superPartition.R]: `seq_along()` is usually a better choice for creating ranges based on length as it avoids bugs related to empty objects

File: R/superPartition.R
Comment (malcolmbarrett) [inline R/superPartition.R:80]: I don't think length will ever return a non-numeric value.

Comment (malcolmbarrett) [inline R/superPartition.R:80]:
```suggestion
    return_mods <- length(modules)
```

File: R/superPartition.R
Comment (malcolmbarrett) [inline R/superPartition.R:84]:
```suggestion
      col_full <- length(unlist(modules[i]))
```

File: R/superPartition.R
Comment (malcolmbarrett) [inline R/superPartition.R]: You almost never want `@import`. Use namespacing, e.g. `tibble::tibble()`, instead, please.

File: R/superPartition.R
Comment (malcolmbarrett) [inline R/superPartition.R]:
```suggestion
    message(paste0("Using `partition()` since there are < ", cluster_size, "features."))
```

File: R/superPartition.R
Comment (malcolmbarrett) [inline R/superPartition.R]:
```suggestion
  i <- j <- k <- l <- 0
```

File: R/superPartition.R
Comment (malcolmbarrett) [inline R/superPartition.R]:
```suggestion
      for (j in seq_along(col_small)) {
```

File: R/superPartition.R
Comment (malcolmbarrett) [inline R/superPartition.R]: This doesn't appear to be necessary. Do you mean to assign another object to `return_mods`?

File: R/superPartition.R
Comment (malcolmbarrett) [inline R/superPartition.R]:
```suggestion
      for (l in seq_along(cs)) {
```

File: R/superPartition.R
Comment (malcolmbarrett) [inline R/superPartition.R]: What warnings are you suppressing here?

File: R/superPartition.R
Comment (malcolmbarrett) [inline R/superPartition.R]:
```suggestion
    for (k in seq_along(mod_rows)) {
```

File: R/superPartition.R
Comment (malcolmbarrett) [inline R/superPartition.R]: Same question as above

**Review by malcolmbarrett — CHANGES_REQUESTED** [review]
Thanks for updating those items! If you could please make these changes, I'll then take a deeper dive into check and the tests

File: R/superPartition.R
Comment (katelynqueen98) [reply R/superPartition.R:80]: This is done to create a blank numeric vector of the appropriate length to then store variables in, not to store the length of the modules object.

File: R/superPartition.R
Comment (katelynqueen98) [reply R/superPartition.R]: You had asked to remove explicit return() statements, which is what used to be here. If I remove this line, then nothing returns from the function.

File: R/superPartition.R
Comment (katelynqueen98) [reply R/superPartition.R]: I get a warning about the new argument not matching the length of the old argument (ie part_master$mapping_key$indices) even though they are both lists of the same length.

File: R/superPartition.R
Comment (malcolmbarrett) [reply R/superPartition.R:80]: Oh, I misread this. I see now. Thank you for clarifying

File: R/superPartition.R
Comment (malcolmbarrett) [reply R/superPartition.R]: That doesn't seem right. `return_mods` alone should return `return_mods`. Can you investigate a bit more?

File: R/superPartition.R
Comment (malcolmbarrett) [reply R/superPartition.R]: hmm ok, I'll look into this. Generally, `suppressWarnings()` gives me itchy feelings, so I'd like to confirm it's necessary. But I'll handle that before merging.

File: R/superPartition.R
Comment (malcolmbarrett) [reply R/superPartition.R]: (I'll also look into this one)

**Comment by malcolmbarrett** [comment]
Note to self: update GitHub action and other CI files to check before merging

File: R/superPartition.R
Comment (katelynqueen98) [reply R/superPartition.R]: Just fixed this!

File: R/superPartition.R
Comment (katelynqueen98) [reply R/superPartition.R]: I don't know why it didn't occur to me to just put return_mods... this is updated now and works correctly!

File: R/superPartition.R
Comment (malcolmbarrett) [reply R/superPartition.R]: Thanks!

**Comment by malcolmbarrett** [comment]
@gvegayon Do you know who ended up with access to Appveyor for the USC repos? I can't access this job to disable it (it's wayyyy old!)

@katelynqueen98 I've made some updates for the CI infrastructure, so please pull those changes if you need to update anything on your end. I also identified the source of the vignette error, which is some problem with the custom director example. I've skipped that test and uneval'd the code chunk so R CMD check will assess the changes in this PR. I'll file an issue to address before we submit this release to CRAN.

File: R/superPartition.R
Comment (malcolmbarrett) [reply R/superPartition.R]: Ok, so this is NOT a warning about `part_master$mapping_key$indices` matching the result of `full_data_col_numbers()`. The warning(s) I see in running the example in the test is:

```
In col_full[j] <- grep(col_small[j], colnames(full_data)) :
  number of items to replace is not a multiple of replacement length
```

It looks like you are trying to find the index of the column name in `col_small[j]`, right? The problem is more than one is being matched in `grep()` (here, `V10` is also being matched because the column name is `V1`, so two indices are being returned). 

I think a better version is matching the name exactly, then using which: `which(colnames(full_data) == col_small[j])`. I bet you could also tweak the regex in the grep() call to only match the name exactly. I don't know which is the faster approach off the top of my head.

This is why suppressing warnings should make you feel itchy! This warning was real.

File: R/superPartition.R
Comment (malcolmbarrett) [inline R/superPartition.R]:
```suggestion
    unique_vals   <- unique(master_cluster$cluster)
```

you're using partial matching here but that is also a common source of bugs

**Comment by malcolmbarrett** [comment]
Alright, a couple more things. Again, be sure to pull as I made some changes to pass R CMD check and set up new Action workflow files.

Also, in playing with this directly, I think you should make the messages and progress bar opt-out, particularly since they are not common in the package. Maybe add a set of `verbose` and `progress_bar` arguments with default `TRUE`? Also, do you think the progress bar is needed for this function based on your experience with it so far? One thing we could do is make that off by default and downgrade the progress package to a suggested package.

**Comment by malcolmbarrett** [comment]
By the way, we're getting close to ready to merge! R CMD Check is passing. You can ignore the appveyor check as that is out of date. I just don't have access to it to disable it.

**Comment by katelynqueen98** [comment]
I think the progress_bar and messages are nice since this is designed for large datasets that may take a significant amount of time to run (~5 hrs for 60,000 features based on testing). But I agree that everyone may not want those, so I added the `verbose` and `progress_bar` options like you suggested.

Thank you for all of your help on this, I really appreciate it!

**Comment by malcolmbarrett** [comment]
My pleasure. That sounds reasonable. I think if you address my other few comments around warnings and partial matching we should be good to merge

**Comment by katelynqueen98** [comment]
I can spend some time tomorrow seeing if I can find a solution for the helper function that doesn't create warnings! I may just be missing it, but I'm not seeing any comments/requested changes around partial matches. Can you clarify what you mean?

**Review by malcolmbarrett — CHANGES_REQUESTED** [review]

**Comment by malcolmbarrett** [comment]
Can you see them now? I didn't submit the second review for requested changes

File: R/superPartition.R
Comment (malcolmbarrett) [reply R/superPartition.R:176]: I bet you could also tweak the regex in the `grep()` call to only match the name exactly. I don't know which is the faster approach off the top of my head

File: R/superPartition.R
Comment (katelynqueen98) [reply R/superPartition.R]: This was a typo! It's fixed in the latest commit

File: R/superPartition.R
Comment (katelynqueen98) [reply R/superPartition.R]: Ah okay that makes sense! I fixed this using `which`, since which is faster for exact matches than grep. I removed the suppressWarnings as well.

Additionally, this made me realize there may be problems if the original data has column names that contain the reduced variable naming string (parameter x). I've added a check at the beginning which requires that no column names contain x:
`if(any(grepl(x, colnames(full_data)))) stop(paste0("The prefix for new variable names, ", x, ", is contained within existing data column names. Please choose a different prefix to avoid errors."))`

I also added an additional set of tests for the errors.

**Review by malcolmbarrett — APPROVED** [review]

**Comment by malcolmbarrett** [comment]
Thanks for your work on this!

**Pull request merged**
